llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -18455,7 +18455,45 @@ SDValue DAGCombiner::visitFP_ROUND(SDNode *N) {
   return SDValue();
 }
 
+// Eliminate a floating-point widening of a narrowed value if the fast math
+// flags allow it.
+static SDValue eliminateFPCastPair(SDNode *N) {
+  SDValue N0 = N->getOperand(0);
+  EVT VT = N->getValueType(0);
+
+  unsigned NarrowingOp;
+  switch (N->getOpcode()) {
+  case ISD::FP16_TO_FP:
+    NarrowingOp = ISD::FP_TO_FP16;
+    break;
+  case ISD::BF16_TO_FP:
+    NarrowingOp = ISD::FP_TO_BF16;
+    break;
+  case ISD::FP_EXTEND:
+    NarrowingOp = ISD::FP_ROUND;
+    break;
+  default:
+    llvm_unreachable("Expected widening FP cast");
+  }
```

Comment:
This should not be a switch/assign/break. Surely there's a utility function for this somewhere already?

Reply:
You could change each call site to pass in `NarrowingOp` as an argument.

Reply:
That then means `visitFP16_TO_FP` needs that same `switch` since despite the name, it is not only called for `ISD::FP16_TO_FP`. That doesn't get rid of the switch, that just moves it.